doc/migration: add sections for v0.8.x and v0.9.x (#1891)
operator-framework/operator-sdk: doc/migration/version-upgrade-guide.md

- Initialize a new `go.mod` file by running `go mod init`.
- Append the following to the end of your `go.mod` file to pin the operator-sdk and other upstream dependencies to the required versions.
```
// Pinned to kubernetes-1.13.1
replace (
	k8s.io/api => k8s.io/api v0.0.0-20181213150558-05914d821849
	k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.0.0-20181213153335-0fe22c71c476
	k8s.io/apimachinery => k8s.io/apimachinery v0.0.0-20181127025237-2b1284ed4c93
	k8s.io/client-go => k8s.io/client-go v0.0.0-20181213151034-8d9ed539ba31
)

replace (
	github.com/coreos/prometheus-operator => github.com/coreos/prometheus-operator v0.29.0
	k8s.io/code-generator => k8s.io/code-generator v0.0.0-20181117043124-c2090bec4d9b
	k8s.io/kube-openapi => k8s.io/kube-openapi v0.0.0-20180711000925-0cf8f7e6ed1d
	sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.1.10
	sigs.k8s.io/controller-tools => sigs.k8s.io/controller-tools v0.1.11-0.20190411181648-9d55346c2bde
)

replace github.com/operator-framework/operator-sdk => github.com/operator-framework/operator-sdk v0.8.2
```

> - Run `go mod tidy` to clean up the `go.mod` file. 

Reviewer: WDYT about we tell to the user's use the make commands in order to add use the GOPROXY (https://github.com/operator-framework/operator-sdk/blob/8bdd7c5bd54b310514059eaaa96c47af2ba6b0c2/Makefile#L31)
```suggestion
- Run `make tidy` to clean up the `go.mod` file. 
```

Reviewer 2: Users don't have a Makefile in their projects, so `make tidy` wouldn't apply to them. That's just a convenience make target we have for the SDK repo.

    version = "=v0.7.1"
  ```

## `v0.8.x`

The SDK version `v0.8.x` supports scaffolding projects to use Go modules by default. It is recommended that you migrate your operator project to use modules for dependency management, however you can choose to keep using `dep`. The upgrade steps for both are outlined below:

### `dep`

- Update the SDK constraint in `Gopkg.toml` to version `v0.8.2`.
  ```TOML
  [[constraint]]
    name = "github.com/operator-framework/operator-sdk"
    version = "=v0.8.2"
  ```
- Pin the controller-tools dependency to the following revision. See the release notes or [#1278](https://github.com/operator-framework/operator-sdk/pull/1278/) for why this is needed.
  ```TOML
  [[override]]
    name = "sigs.k8s.io/controller-tools"
    revision = "9d55346c2bde73fb3326ac22eac2e5210a730207"
  ```
- Run `dep ensure` to update the vendor directory.

### `modules`

To get familiar with Go modules read the [modules wiki][modules-wiki]. In particular the section on [migrating to modules][migrating-to-modules].

> - Ensure that you have Go 1.12+ and [Mercurial][mercurial] 3.9+ installed.

Reviewer: ```suggestion
- Ensure that you have Go 1.12+ (I do not think that The Mercurial is required at all)
```

Reviewer: Also, is required to:
-  create the `group.go`
- Ensure that ObjectStruct has

```
// +k8s:openapi-gen=true
// +kubebuilder:subresource:status// 
```

Author: Adding `group.go` isn't really necessary since it's not a breaking change. It only suppresses a warning from gengo when running the gen-code commands.

But you're right on the second point: I should highlight the need for `+kubebuilder` tags now that CRD manifests are overwritten in `v0.8.x` and will remove any manual edits made by users to the CRD.
From the CHANGELOG:
```
- Go operator CRDs are overwritten when being regenerated by operator-sdk generate openapi. 
Users can now rely on +kubebuilder annotations in their API code, which provide access to most OpenAPIv3
 validation properties (the full set will be supported in the near future, see this PR) and other CRD fields. (#1278)
```
This probably should have been marked as breaking change to begin with. Well spotted.

Author: Also I can't recall if I had to setup mercurial when I used Go modules but it's definitely part of our pre-reqs:
https://github.com/operator-framework/operator-sdk#prerequisites

Although it might be one of git or mercurial:
https://github.com/golang/go/wiki/GoGetTools

I'll need to double check and remove it from the SDK's pre-reqs if that's not the case.

Reviewer 2: I'm fairly sure mercurial was/is required. I'm pretty sure some transitive dependency somewhere comes from a mercurial repo.

Also, just be careful about _which_ kubebuilder tags are mentioned about being needed/supported. We won't switch to controller-tools 0.2.x until the next release, so tags like `+kubebuilder:subresource:status` that were added in controller-tools 0.2.0 should not be mentioned.

Author: Yeah I've linked to the legacy kubebuilder docs to show that not all marker comments are supported for the controller-tools version tied SDK release v0.8.2.

But `// +kubebuilder:subresource:status` specifically is supported pre controller-tools v0.2.0 so I've kept that in the example.

Reviewer: HI @joelanford, @hasbro17  I have no mercurial installed locally and all are working. 

```
$ mercurial
-bash: mercurial: command not found
```

Reviewer 2: @camilamacedo86 The mercurial command is `hg`, so check if you have that.

Here's the original issue: https://github.com/operator-framework/operator-sdk/issues/1681

Looks like it affected 0.9 at least, not sure about 0.8 or 0.10.

Reviewer: I lot of shame now :-P Yes, I have it! All explained tks for the patient with me.

- Initialize a new `go.mod` file by running `go mod init`.
- Append the following to the end of your `go.mod` file to pin the operator-sdk and other upstream dependencies to the required versions.
```
// Pinned to kubernetes-1.13.1
replace (
	k8s.io/api => k8s.io/api v0.0.0-20181213150558-05914d821849
	k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.0.0-20181213153335-0fe22c71c476
	k8s.io/apimachinery => k8s.io/apimachinery v0.0.0-20181127025237-2b1284ed4c93
	k8s.io/client-go => k8s.io/client-go v0.0.0-20181213151034-8d9ed539ba31
)

replace (
	github.com/coreos/prometheus-operator => github.com/coreos/prometheus-operator v0.29.0
	k8s.io/code-generator => k8s.io/code-generator v0.0.0-20181117043124-c2090bec4d9b
	k8s.io/kube-openapi => k8s.io/kube-openapi v0.0.0-20180711000925-0cf8f7e6ed1d
	sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.1.10
	sigs.k8s.io/controller-tools => sigs.k8s.io/controller-tools v0.1.11-0.20190411181648-9d55346c2bde
)

replace github.com/operator-framework/operator-sdk => github.com/operator-framework/operator-sdk v0.8.2

> ```

Reviewer: I think here we can add to remove the Go dep files.

Author: Can you outline why exactly?

Right now that's the last step once they've confirmed they can build the project with modules enabled.

Plus I think running `go mod tidy` shouldn't be affected by the presence of Gopkg.toml or the vendor directory.

Reviewer: You are correct after I recheck it . Please ignore this comment.

- Run `go mod tidy` to clean up the `go.mod` file. 
  - In case of any go module loading errors, consult the default [`v0.8.2` go.mod dependencies][v0.8.2-go-mod] scaffolded by the operator-sdk to resolve any differences. You can also view this file by scaffolding a new project with operator-sdk `v0.8.2`.
- Ensure that you can build the project with `operator-sdk build`
- Finally remove `Gopkg.lock`, `Gopkg.toml` and the vendor directory.

## `v0.9.x`

> - The function `ExposeMetricsPort()` has been replaced with `CreateMetricsService()` [#1560](https://github.com/operator-framework/operator-sdk/pull/1560). 

Reviewer: I think here we have more changes than that. See [here](https://github.com/operator-framework/operator-sdk-samples/pull/72/commits/f66246e19a1855c34f94c1417559934665110beb#diff-b4731807d0f451420df4f54133ce69f2R7) this file migrate to 0.10 at least. But is required to create a new project with the 0.9.0 and compare. I am not sure if all changes here were introduced in the 0.10 or 0.9.

Author: @camilamacedo86 `ExposeMetricsPort()`==>`CreateMetricsService()` is the only breaking change in `v0.8.x` to `v0.9.x`. You can compare the changes here for the scaffolded main.go:
https://github.com/operator-framework/operator-sdk/compare/v0.8.x...v0.9.x#diff-17dd005b8bf50b6dbab49cc2758e0498R162

The PR that you've opened is taking the sample project from `v0.6.0` to `v0.10.0`.
Plus the other changes to the scaffolded cmd/main.go such as [`serveCRMetrics()`](https://github.com/operator-framework/operator-sdk/blob/master/internal/pkg/scaffold/cmd.go#L152-L154) are not breaking changes that are needed to upgrade the versions. It's just an additional feature so I'm not outlining everything that needs to be changed. Just the minimum to avoid breaking changes and dependency updates.

Reviewer: Closing as solved 👍

  [[override]]
    name = "k8s.io/client-go"
    # revision for tag "kubernetes-1.13.4"
    revision = "b40b2a5939e43f7ffe0028ad67586b7ce50bb675"
  [[override]]
    name = "github.com/coreos/prometheus-operator"
    version = "=v0.29.0"
  [[override]]
    name = "sigs.k8s.io/controller-runtime"
    version = "=v0.1.12"
  [[constraint]]
    name = "github.com/operator-framework/operator-sdk"
    version = "=v0.9.0"
  ```
- Append the contraint for `k8s.io/kube-state-metrics`.
  ```TOML
  [[override]]
    name = "k8s.io/kube-state-metrics"
    version = "v1.6.0"
  ```
- Run `dep ensure` to update the vendor directory.

### modules

- Update the `replace` directives in your `go.mod` file for the SDK, kubernetes, controller-runtime and kube-state metrics dependencies to the following versions.
  ```
  // Pinned to kubernetes-1.13.4
  replace (
    k8s.io/api => k8s.io/api v0.0.0-20190222213804-5cb15d344471
    k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.0.0-20190228180357-d002e88f6236
    k8s.io/apimachinery => k8s.io/apimachinery v0.0.0-20190221213512-86fb29eff628
    k8s.io/client-go => k8s.io/client-go v0.0.0-20190228174230-b40b2a5939e4
  )

>   replace (

Reviewer: I think it need be updated as here; https://github.com/operator-framework/operator-sdk/pull/1899

Author: My first thought is probably not.
There are other ways to solve this problem, e.g using the proxy cache by setting `GOPROXY=https://proxy.golang.org`.

I don't want to add it to the guide since this isn't an issue specifically with the SDK and something that's easily if you run into the issue.

Although it looks like https://status.apache.org/ is down quite often so maybe if we merge #1899 I can do a follow up.

	k8s.io/client-go => k8s.io/client-go v0.0.0-20181213151034-8d9ed539ba31
)

replace (
	github.com/coreos/prometheus-operator => github.com/coreos/prometheus-operator v0.29.0
	k8s.io/code-generator => k8s.io/code-generator v0.0.0-20181117043124-c2090bec4d9b
	k8s.io/kube-openapi => k8s.io/kube-openapi v0.0.0-20180711000925-0cf8f7e6ed1d
	sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.1.10
	sigs.k8s.io/controller-tools => sigs.k8s.io/controller-tools v0.1.11-0.20190411181648-9d55346c2bde
)

replace github.com/operator-framework/operator-sdk => github.com/operator-framework/operator-sdk v0.8.2
```
- Run `go mod tidy` to clean up the `go.mod` file. 
  - In case of any go module loading errors, consult the default [`v0.8.2` go.mod dependencies][v0.8.2-go-mod] scaffolded by the operator-sdk to resolve any differences. You can also view this file by scaffolding a new project with operator-sdk `v0.8.2`.
- Ensure that you can build the project with `operator-sdk build`
- Finally remove `Gopkg.lock`, `Gopkg.toml` and the vendor directory.

### Breaking changes

Upon updating the project to `v0.8.2` the following breaking changes apply:

> - On running the command `operator-sdk generate openapi`, the CRD manifests at `deploy/crds/<group>_<version>_<kind>.crd` for all API types will now be regenerated based on their source files `pkg/apis/..._types.go`. So if you have made any manual edits to the default generated CRD manifest, e.g manually written the validation block or specified the naming(`spec.names`), then that information be overwritten when the CRD is regenerated. 

Reviewer 2: Nit: add space between `naming` and `` (`spec.names`) ``
```suggestion
- On running the command `operator-sdk generate openapi`, the CRD manifests at `deploy/crds/<group>_<version>_<kind>.crd` for all API types will now be regenerated based on their source files `pkg/apis/..._types.go`. So if you have made any manual edits to the default generated CRD manifest, e.g manually written the validation block or specified the naming (`spec.names`), then that information be overwritten when the CRD is regenerated. 
```
